serial-settings/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+#![no_std]
+
+use core::fmt::Write;
+use miniconf::JsonCoreSlash;
+use embedded_storage::nor_flash::NorFlash;
+use serde::Serialize;
+
+static OUTPUT_BUFFER: bbqueue::BBBuffer<512> = bbqueue::BBBuffer::new();
+
+struct Context<Settings: SerialSettings, Flash: NorFlash + 'static> {
+    output: OutputBuffer,
+    flash: Flash,
+    settings: Settings,
+}
+
+impl<Settings: SerialSettings, Flash: NorFlash + 'static> Context<Settings, Flash> {
+    fn save(&mut self) {
+        self.flash.erase(0, self.flash.capacity() as u32).unwrap();
+        // TODO: Use a user provided buffer.
+        let mut buf = [0; 512];
+        let serialized = postcard::to_slice(&self.settings, &mut buf).unwrap();
+        self.flash.write(0, serialized).unwrap();
+    }
+}
+
+pub trait SerialSettings: for<'a> JsonCoreSlash<'a> + Serialize + Clone + 'static {
+    fn reset(&mut self);
+}
+
+fn make_menu<Settings: SerialSettings, Flash: NorFlash + 'static>() -> menu::Menu<'static, Context<Settings, Flash>> {
```

Comment:
I wonder if there's an easy way for us to make the menu options expandable by the user... Perhaps defer to the future

Reply:
Maybe not needed now. All we want to add now is `service` and `dfu` AFAICT.

Reply:
Both `service` and `dfu` are not things we have on stabilizer. We can add them, but resetting to DFU mode is generally an involved process, so I'd imagine it's going to be a function provided on the trait object.

Reply:
I added an abstraction in214d29fce5cf46e5cbbd317e5ac0e0051034937b, but I'm not the biggest fan of it because it means that there's `service` and `dfu` menu options in stabilizer that don't do anything currently (although we could add them).

Ideally, these items wouldn't even exist in the menu if they're not desired, but we can hold off on this for the future